---
title: "GC content and epigenetic ChIP"
author: "Sharonhe"
date: "November 5, 2016"
output: html_document
---

# Dependencies
library(AnnotationHub)
library(BSgenome)
library(rtracklayer)
library(GenomicFeatures)

# Project 1: compute the GC content of promoters in the yeast genome. 

## Get yeast genome
```{r}
library("BSgenome.Scerevisiae.UCSC.sacCer2")
Scerevisiae
```


## Get promoter ranges of yeast genome

```{r}
ahub = AnnotationHub()
qh = query(ahub, c("sacCer2", "genes"))
qh
genes = ahub[["AH7048"]]
prom = promoters(genes)
# There is warning messages here because G ranges complain when it gets like genome indices that are less than zero. This problem can be solved by trimming promoters
prom = trim(prom)
prom
```

## View the promoter

```{r}
promViews = Views(Scerevisiae, prom)
```


## Get GC content in the promoter and make a density plot

```{r}
gcProm = letterFrequency(promViews, "GC", as.prob = TRUE)
plot(density(gcProm))
abline(v = 0.38)
```

# Project 2: Compute the GC content of “chr22” in the “hg19” build of the human genome

## Get hg19 human genome

```{r}
source("https://bioconductor.org/biocLite.R")
biocLite("BSgenome.Hsapiens.UCSC.hg19")
library(BSgenome.Hsapiens.UCSC.hg19)
Hsapiens
```

# count total bases on chr22

```{r}
totalBase <- letterFrequency(Hsapiens$chr22, "A") +
             letterFrequency(Hsapiens$chr22, "C") +
             letterFrequency(Hsapiens$chr22, "G") +
             letterFrequency(Hsapiens$chr22, "T")
```

# count GC bases on chr22  

```{r}
gcBase <- letterFrequency(Hsapiens$chr22, "GC")
```

# calculate GC ratio on chr22

```{r}
gcContent <- gcBase/totalBase
gcContent
```

# Project 3: Compute the mean GC content of H3K27me3 “narrowPeak” regions from Epigenomics Roadmap from the H1 stem cell line on chr 22.

## getting data of H3K27me3 “narrowPeak” regions from Epigenomics Roadmap

```{r}
ah <- AnnotationHub()
H3K27me3_qh <- query(ah, c("H3K27me3", "E003", "narrowPeak"))
H3K27me3_qh
H3K27me3_data <- H3K27me3_qh[["AH29892"]] # retrieve data
```

## get genomic ranges on chr22 from the above epigenetic data 

```{r}
H3K27me3_data.chr22 <- subset(H3K27me3_data, seqnames == "chr22")
```

## get the corresponding DNA sequence

```{r}
library(BSgenome.Hsapiens.UCSC.hg19)
Hsapiens
H3K27me3_data.chr22.seq <- Views(Hsapiens, H3K27me3_data.chr22)
```

# calculate GC content on chr22

```{r}
gcContents <- letterFrequency(H3K27me3_data.chr22.seq, "GC", as.prob = T)
meanGcContents <- mean(gcContents)
meanGcContents
```

# Compute the correlation between GC content and “signalValue” of these regions (on chr22)

```{r}
sigV <- mcols(H3K27me3_data.chr22.seq)$signalValue
cor(sigV, gcContents) # calculate correlation
```


# Project 4: Use "fc.signal" data to comfirm whether “narrowPeak” regions are presumably reflective of a ChIP signal on chr22.(compute their correlation)

## getting fc.signal data

```{r}
H3K27me3_fc <- query(ah, c("H3K27me3", "E003", "fc.signal"))
H3K27me3_fc
H3K27me3_fc.data <- H3K27me3_fc[["AH32033"]]
# fc.signal data represents a vector of fold-change enrichment of ChIP signal over input.
```

## subset data on chr22 from fs.signal data

```{r}
gr.chr22 <- GRanges(seqnames = "chr22", ranges = IRanges(start =1, end = 51304566))
H3K27me3_fc.rel <- import(H3K27me3_fc.data, which =gr.chr22, as = "Rle")
H3K27me3_fc.rel.chr22 <- H3K27me3_fc.rel$chr22
```

## view fc.signal data

```{r}
fc.signal.chr22 <- Views(H3K27me3_fc.rel.chr22, start = start(H3K27me3_data.chr22), end = end(H3K27me3_data.chr22))
```

## calculate mean values of fc.signal across each region on chr22

```{r}
fc.signal.mean <- mean(fc.signal.chr22)
```

## calculate correlation between fc.singal and signal value
```{r}
cor(fc.signal.mean, sigV)
```


## Count the bases on chr22 that have an fc.signal greater than or equal to 1.

```{r}
sum(H3K27me3_fc.rel.chr22 >= 1)
```

# Project 5: Identify regions of the genome which gain H3K27me3 upon differentiation. 

The H1 stem cell line is an embryonic stem cell line, a so-called pluripotent cell. Many epigenetic marks change upon differentiation. We want to identify regions of the genome which gain H3K27me3 upon differentiation. These are regions which have a higher signal in E055 than in E003. Identify the regions on chr22 where the fc.signal in E003 is 0.5 or lower and the fc.signal in E055 is 2 or higher.

## get fc.signal in E055

```{r}
H3K27me3_fc_E005 = query(ah, c("H3K27me3", "E005", "fc.signal"))
H3K27me3_fc_E005
H3K27me3_fc_E005_data = H3K27me3_fc_E005[[1]]
```

## get subset data on chr22 in E055

```{r}
# This is the method of importing a bigwig file with rtracklayer package
gr.chr22 <- GRanges(seqnames = "chr22", ranges = IRanges(start =1, end = 51304566))
H3K27me3_fc.rel_E055 <- import(H3K27me3_fc_E055_data, which =gr.chr22, as = "Rle") 
H3K27me3_fc.rel_E055.chr22 <- H3K27me3_fc.rel_E055$chr22
```

## identify region

```{r}
region_E003 <- slice(H3K27me3_fc.rel.chr22, upper = 0.5)
region_E055 <- slice(H3K27me3_fc.rel_E055.chr22, lower = 2)

# in order to use intersect function, we have to convert the Views to IRanges or GRanges first. 
region_E003 <- as(region_E003, "IRanges") 
region_E055 <- as(region_E055, "IRanges")

inter_region <- intersect(region_E003, region_E055)
sum(width(inter_region))
```

# Project 6: Compute the average observed-to-expected ratio of CpG dinucleotides for CpG Islands on chromosome 22 in hg19

CpG Islands are dense clusters of CpGs. The classic definition of a CpG Island compares the observed to the expected frequencies of CpG dinucleotides as well as the GC content.

Specifically, the observed CpG frequency is just the number of “CG” dinucleotides in a region divided by the length of the region. The expected CpG frequency is defined as the frequency of C multiplied by the frequency of G divided by the length of the region.

## get hg19 CpG dataset

```{r}
CpGIsland <- query(ah, c("hg19", "CpG Islands"))
CpGIsland_data <- CpGIsland[["AH5086"]]
```

## subset and view genome data on chr22

```{r}
CpGIsland_data.chr22 <- subset(CpGIsland_data, seqnames == "chr22")
library(BSgenome.Hsapiens.UCSC.hg19)
Hsapiens
CpGIsland_data.chr22.vi <- Views(Hsapiens, CpGIsland_data.chr22)
```

## calculate observed GC bases (the number of “CG” dinucleotides in a region)

```{r}
region_length <- width(CpGIsland_data.chr22.vi)
observed_gcBase <- dinucleotideFrequency(CpGIsland_data.chr22.vi)[,7]/region_length
```


# Project 7: TATA box in human genome

## Compute the number of TATA boxes on chr 22 of build hg19 of the human genome

A TATA box is a DNA element of the form “TATAAA”. Around 25% of genes should have a TATA box in their promoter. In this project we need to remember to search both forward and reverse strands.

```{r}
library(BSgenome.Hsapiens.UCSC.hg19)
TATA_boxes <- countPattern("TATAAA", Hsapiens$chr22) + countPattern("TATAAA", reverseComplement(Hsapiens$chr22))
TATA_boxes
```

## Compute how many promoters of transcripts on chromosome 22 containing a coding sequence, contains a TATA box on the same strand as the transcript

Use the TxDb.Hsapiens.UCSC.hg19.knownGene package to define transcripts and coding sequence. Here, we defined a promoter to be 900bp upstream and 100bp downstream of the transcription start site.

```{r}
library(TxDb.Hsapiens.UCSC.hg19.knownGene)
txdb <- TxDb.Hsapiens.UCSC.hg19.knownGene

# The following is the method of subsetting a txdb database

# get transcript on chr22
gr <- GRanges(seqnames = "chr22", ranges = IRanges(start = 1, end = 52330658))
gr.trans.chr22 <- subsetByOverlaps(transcripts(txdb), gr, ignore.strand = TRUE)

# get promoters on chr22
proms <- promoters(gr.trans.chr22, upstream = 900, downstream = 100)

# get coding sequence on chr22
cdseq <- subsetByOverlaps(cdsby(txdb), gr, ignore.strand = TRUE)

# get promoters that have coding sequence
proms_cds <- findOverlaps(proms, cdseq)
unique(queryHits(proms_cds))

# count
count = 0
for (i in unique(queryHits(proms_cds))){
  proms_cds_vi <- Views(Hsapiens, proms[i])
  count = count + vcountPattern("TATAAA", DNAStringSet(proms_cds_vi))
}
count

```



sessioninfor()
